WalletWasabi/WabiSabi/Client/AliceClient.cs

```diff
@@ -0,0 +1,106 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Bases;
+using WalletWasabi.Crypto.ZeroKnowledge;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Models;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class AliceClient : BackgroundService, IAsyncDisposable
+	{
+		public Guid AliceId { get; }
+		public Guid RoundId { get; }
+		private ArenaClient ArenaClient { get; }
+		private IEnumerable<Coin> Coins { get; }
+
+		public AliceClient(Guid aliceId, Guid roundId, ArenaClient arenaClient, IEnumerable<Coin> coins)
+		{
+			AliceId = aliceId;
+			RoundId = roundId;
+			ArenaClient = arenaClient;
+			Coins = coins;
+		}
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			try
+			{
+				do
+				{
+					await Task.Delay(TimeSpan.FromSeconds(30), stoppingToken).ConfigureAwait(false);
+				}
+				while (!await ConfirmConnectionAsync().ConfigureAwait(false));
+			}
+			catch (Exception ex)
+			{
+				Logger.LogError($"Round ({RoundId}), Alice ({AliceId}): ConfirmConnection failed, reason: '{ex}'");
```

Comment:
There is a LogError method that receives the exception and formats the message for use, it is basically the same that you do here but is just to the record. Anyway, there are too many ways to log things, just take a look at the Yahia's `LoggerTools`. IMO, this is okay as it is here but we shouldn't invest a lot of time on logging right now because we have to improve the way we are logging.

Reply:
agreed, i have many thoughts about logging but we should do that after we have various components in place

Reply:
It is not the same. `RoundId` and `AliceId` seem to be useful for debugging. I won't change this as we agreed here. More interesting question, what shall we do if there is an unexpected failure, we will never get to the confirmation phase.

Reply:
I think all exceptions should be either passed through or in this case rethrown after logging. Whatever the per-round client ends up as, it should be waiting on `Alice`s to get the inputs confirmed and should be responsible for collecting and handling errors for them. I can't think of any errors it would make sense for `Alice` to try to handle, transport errors should not happen until the transport layer considers them final/unrecoverable.

Reply:
This is a separate task running in the ThreadPool. The exception will end up nowhere - to be precise the exception will be thrown where the Task awaited - which is at StopAsync inside DisposeAsync function. 
We could try to RemoveInputAsync and then indicate the failure with a property, event, or whatever depends on the usage. I started to add tests where this will turn out.

Reply:
If input confirmation fails the input will time out, I don't think it's necessary to deregister it, it will not affect the round since it will just not get confirmed, and I can't think of a situation where connection confirmation falis but deregistration would work apart from ephemeral transport errors which should not be visible at this abstraction layer.

The correct behaviour would be to propagate the error, if for technical reasons we can't do that by having the supervising per-round client await on the task orthogonally to the thread pool (I'm not yet familiar with those APIs, I will study up on them now) considerations, what's an alternative way to convey successful confirmation or an exception? Can the threadpool level task wrap an inner task that can be awaited by the supervising client?

Reply:
I changed the structure here. There will be some object that will arrange Alice and Bob client like for example CoinJoinClient. So Alice does not need to manage a dedicated loop. It is enough to have it in one method.

Reply:
With this, the error propagation is also straightforward.